### PR TITLE
ARROW-7366: [C++][Dataset] Use PartitionSchemeDiscovery in DataSourceDiscovery

### PR DESCRIPTION
**Pull request by bkietz**

### CODE REVIEW

**Comment by fsaintjacques** [comment]
Needs a rebase.

File: r/src/dataset.cpp [hunk 1]
```diff
@@ -22,20 +22,26 @@
...
     const std::shared_ptr<fs::FileSystem>& fs,
-    const std::shared_ptr<fs::Selector>& selector) {
+    const std::shared_ptr<fs::Selector>& selector,
+    const std::shared_ptr<ds::PartitionScheme>& partition_scheme) {
   // TODO(npr): add format as an argument, don't hard-code Parquet
   auto format = std::make_shared<ds::ParquetFileFormat>();
+
   // TODO(fsaintjacques): Make options configurable
   auto options = ds::FileSystemDiscoveryOptions{};
+  if (partition_scheme != nullptr) {
```
Comment (nealrichardson) [inline r/src/dataset.cpp]:
```suggestion
  if (!Rf_isNull(partition_scheme)) {
```

Comment (nealrichardson) [reply r/src/dataset.cpp]: No this was wrong, need two versions of the function, one with and one without partition_scheme arg. See r/src/array.cpp for example.

File: r/R/dataset.R [hunk 2]
```diff
@@ -123,7 +123,7 @@ names.Dataset <- function(x) names(x$schema)
 #'
 #' - `$Inspect()`: Walks the files in the directory and returns a common [Schema]
 #' - `$SetPartitionScheme(part)`: Takes a [PartitionScheme]
```
Comment (bkietz) [inline r/R/dataset.R]: delete

File: r/src/dataset.cpp (hunk 1 above)
Comment (bkietz) [reply r/src/dataset.cpp]: Thanks, that's working

File: cpp/src/arrow/dataset/dataset_test.cc [hunk 3]
```diff
@@ -335,17 +332,21 @@ TEST_F(TestEndToEnd, EndToEndSingleSource) {
   // - "/2019/01 -> {"year": 2019, "month": 1}
   // - "/2019/01/CA -> {"year": 2019, "month": 1, "country": "CA"}
   // - "/2019/01/CA/a_file.json -> {"year": 2019, "month": 1, "country": "CA"}
-  auto partition_scheme = std::make_shared<SchemaPartitionScheme>(partition_schema);
-  ASSERT_OK(discovery->SetPartitionScheme(partition_scheme));
+  options.partition_scheme =
```
Comment (fsaintjacques) [inline cpp/src/arrow/dataset/dataset_test.cc]: Update documentation to reflect that the schema will be guessed.

File: cpp/src/arrow/dataset/discovery.h [hunk 4]
```diff
@@ -59,21 +61,13 @@ class ARROW_DS_EXPORT DataSourceDiscovery {
...
-    return Status::OK();
-  }
-
-  const PartitionSchemePtr& partition_scheme() const { return partition_scheme_; }
-  Status SetPartitionScheme(PartitionSchemePtr partition_scheme) {
-    partition_scheme_ = partition_scheme;
-    return Status::OK();
-  }
+  /// \brief Create a DataSource using an inspected schema.
+  Result<DataSourcePtr> Finish();
```
Comment (fsaintjacques) [inline cpp/src/arrow/dataset/discovery.h]: Use an optional parameter and merge the previous definitions. See `discovery.cc` about the behavior of this function, e.g. validation of explicit schema.

File: cpp/src/arrow/dataset/discovery.h [hunk 5]
```diff
@@ -85,12 +79,20 @@ class ARROW_DS_EXPORT DataSourceDiscovery {
...
 struct FileSystemDiscoveryOptions {
+  // Either an explicit PartitionScheme or a PartitionSchemeDiscovery to discover one.
+  //
+  // If a discovery is provided, it will be used to infer a schema for partition fields
+  // based on file and directory paths then construct a PartitionScheme. The default is
+  // to discover a partition scheme assuming hive style directory naming.
+  //
+  // The resulting partition scheme will be applied to discovered files and the resulting
+  // partition information embedded in the DataSource.
+  PartitionSchemeOrDiscovery partition_scheme{HivePartitionScheme::MakeDiscovery()};
```
Comment (fsaintjacques) [inline cpp/src/arrow/dataset/discovery.h]: This has the default of unpleasant surprises if a user doesn't expect it, the default should be none.

File: cpp/src/arrow/dataset/discovery.cc [hunk 6]
```diff
@@ -119,24 +114,53 @@ Result<DataSourceDiscoveryPtr> FileSystemDataSourceDiscovery::Make(
...
 }
 
-Result<DataSourcePtr> FileSystemDataSourceDiscovery::Finish() {
-  ExpressionVector partitions(files_.size(), scalar(true));
+Result<DataSourcePtr> DataSourceDiscovery::Finish() {
+  ARROW_ASSIGN_OR_RAISE(auto schema, Inspect());
+  return Finish(schema);
+}
+
+Result<DataSourcePtr> FileSystemDataSourceDiscovery::Finish(
```
Comment (fsaintjacques) [inline cpp/src/arrow/dataset/discovery.cc]: The schema handling part of this body should be:

- If a schema is provided, validate it against `Inspect` result.
- If a schema is not provided, use `Inspect` result as the schema.

Think about other DataSource, e.g. OdbcData, how should it validate. The struct options in discovery is class specific, so maybe with functions:

```c++
// Validate infer against provided or use infered schema.
Finish(const std::shared_ptr<Schema>& schema = nullptr);
// Take as-is with no validation, might blow up at runtime.
FinishUnsafe(const std::shared_ptr<Schema>& schema);
```

File: cpp/src/arrow/dataset/discovery.cc [hunk 7]
```diff
@@ -146,18 +170,18 @@ Result<DataSourcePtr> FileSystemDataSourceDiscovery::Finish() {
...
+        auto maybe_partition = partition_scheme->Parse(segments.back(), segment_index);
 
         partitions[ref.i] = std::move(maybe_partition).ValueOr(scalar(true));
       }
     }
     return Status::OK();
   };
 
-  RETURN_NOT_OK(forest.Visit(apply_partition_scheme));
+  RETURN_NOT_OK(forest_.Visit(apply_partition_scheme));
```
Comment (fsaintjacques) [inline cpp/src/arrow/dataset/discovery.cc]: I'm not sure how you want to deal here, either nullptr or NoopScheme. I tend to prefer nullptr, but you don't.

File: python/pyarrow/_dataset.pyx [hunk 8]
```diff
@@ -186,6 +207,23 @@ cdef class FileSystemDiscoveryOptions:
     cdef inline CFileSystemDiscoveryOptions unwrap(self):
         return self.options
 
+    @property
+    def partition_scheme(self):
+        cdef shared_ptr[CPartitionScheme] s = self.options.partition_scheme.scheme()
+        if s.get() != nullptr:
+            return PartitionScheme.wrap(s)
+        return PartitionSchemeDiscovery.wrap(self.options.partition_scheme.discovery())
```
Comment (fsaintjacques) [inline python/pyarrow/_dataset.pyx]: Return `None` and add a `partition_scheme_discovery` method.

File: python/pyarrow/_dataset.pyx [hunk 9]
```diff
@@ -275,10 +300,13 @@ cdef class DataSourceDiscovery:
             result = self.discovery.Inspect()
         return pyarrow_wrap_schema(GetResultValue(result))
 
-    def finish(self):
-        cdef CResult[shared_ptr[CDataSource]] result
+    def finish(self, Schema schema):
```
Comment (fsaintjacques) [inline python/pyarrow/_dataset.pyx]: Schema should not be mandatory

File: python/pyarrow/includes/libarrow_dataset.pxd [hunk 10]
```diff
@@ -324,20 +327,26 @@ cdef extern from "arrow/dataset/api.h" namespace "arrow::dataset" nogil:
...
 
     cdef cppclass CDataSourceDiscovery "arrow::dataset::DataSourceDiscovery":
         CResult[vector[shared_ptr[CSchema]]] InspectSchemas()
         CResult[shared_ptr[CSchema]] Inspect()
-        CResult[CDataSourcePtr] Finish()
-        shared_ptr[CSchema] schema()
-        CStatus SetSchema(shared_ptr[CSchema])
-        CPartitionSchemePtr partition_scheme()
-        CStatus SetPartitionScheme(CPartitionSchemePtr partition_scheme)
+        CResult[CDataSourcePtr] Finish(shared_ptr[CSchema])
```
Comment (fsaintjacques) [inline python/pyarrow/includes/libarrow_dataset.pxd]: Expose the schemaless version.

**Review by fsaintjacques — CHANGES_REQUESTED** [review]
Looks good in general:
- PartitionScheme/PartitionSchemeDiscovery should be optional, there's a lot of the code that assumes otherwise. Such internal details should not leak in the exposed interface. Python lost the ability to not pass a schema.
- Drop the timestamp conversion stuff from this PR, make a seperate issue.

File: cpp/src/arrow/dataset/discovery.h (hunk 5 above)
Comment (bkietz) [reply cpp/src/arrow/dataset/discovery.h]: I have no opinion on this; I was just following the behavior of `pyarrow.ParquetDataset`. I'll revert it but It seems like a reasonable default for the scripting languages at least (similar to inserting implicit casts). @nealrichardson @kszucs you may consider adding `HivePartitionScheme::MakeDiscovery()` as the default for this field in R/Python

File: cpp/src/arrow/dataset/dataset_test.cc (hunk 3 above)
Comment (bkietz) [reply cpp/src/arrow/dataset/dataset_test.cc]: will do

File: cpp/src/arrow/dataset/discovery.h (hunk 4 above)
Comment (bkietz) [reply cpp/src/arrow/dataset/discovery.h]: alright

File: python/pyarrow/_dataset.pyx (hunk 8 above)
Comment (bkietz) [reply python/pyarrow/_dataset.pyx]: will do

File: cpp/src/arrow/dataset/discovery.cc (hunk 7 above)
Comment (bkietz) [reply cpp/src/arrow/dataset/discovery.cc]: DefaultPartitionScheme is a noop and it's the old default for the partition_scheme field. I'll just re-add that

File: python/pyarrow/includes/libarrow_dataset.pxd (hunk 10 above)
Comment (bkietz) [reply python/pyarrow/includes/libarrow_dataset.pxd]: will do

File: cpp/src/arrow/dataset/discovery.h (hunk 4 above)
Comment (bkietz) [reply cpp/src/arrow/dataset/discovery.h]: Actually I'm not sure I agree with this one; default parameters in a virtual method pass the default case handling down to each implementation. Is there any behavior which should be supported for "Finish without explicit schema" other than "inspect a schema then finish with that"?

**Comment by bkietz** [comment]
@fsaintjacques there *will* always be a partition scheme; the default is `DefaultPartitionScheme` which is a noop and attaches `scalar(true)` to everything. I'll extract the timestamp conversion changes

File: cpp/src/arrow/dataset/discovery.h (hunk 5 above)
Comment (fsaintjacques) [reply cpp/src/arrow/dataset/discovery.h]: Python: batteries included. C++: sharp knives scattered on the floor.

Comment (bkietz) [reply cpp/src/arrow/dataset/discovery.h]: ... and nothing is on by default

File: cpp/src/arrow/dataset/discovery.h (hunk 4 above)
Comment (fsaintjacques) [reply cpp/src/arrow/dataset/discovery.h]: Then let's remove the virtual case for now :)

Comment (fsaintjacques) [reply cpp/src/arrow/dataset/discovery.h]: I think it's an anti-pattern to expect the user to call the Validate method so it doesn't blow up in his face. Anyhow, I'll tackle this in my next PR on DatasetDiscovery as discrepency will likely happen there more than in the DataSourceDiscovery.

**Review by fsaintjacques — APPROVED** [review]

**Pull request closed without merging**